pkg/internal/server/grpc/common_test.go

```diff
@@ -62,12 +66,70 @@ func newTestHarness() *testHarness {
 	}
 }
 
+func newTestHarnessWithClientAPIs() *testHarness {
+	th := newTestHarness()
+
+	th.fakeKubeClient = fake.NewSimpleClientset()
+	th.fakeKubeClient.PrependReactor("create", "tokenreviews", func(action clientgotesting.Action) (handled bool, ret apiruntime.Object, err error) {
+		ca := action.(clientgotesting.CreateAction)
+		trs := ca.GetObject().(*authv1.TokenReview)
+		trs.Status.Authenticated = false
+		if trs.Spec.Token == th.SecurityToken {
+			trs.Status.Authenticated = true
+			trs.Status.Audiences = []string{th.Audience, "other-" + th.Audience}
+		}
+		return true, trs, nil
+	})
+	th.fakeKubeClient.PrependReactor("create", "subjectaccessreviews", func(action clientgotesting.Action) (handled bool, ret apiruntime.Object, err error) {
+		ca := action.(clientgotesting.CreateAction)
+		sar := ca.GetObject().(*v1.SubjectAccessReview)
+		if sar.Spec.ResourceAttributes != nil && sar.Spec.ResourceAttributes.Namespace == th.Namespace {
+			sar.Status.Allowed = true
+		} else {
+			sar.Status.Allowed = false
+			sar.Status.Reason = "namespace mismatch"
+		}
+		return true, sar, nil
+	})
+
+	th.fakeSnapshotClient = fakesnapshot.NewSimpleClientset()
+	th.fakeSnapshotClient.PrependReactor("get", "volumesnapshots", func(action clientgotesting.Action) (handled bool, ret apiruntime.Object, err error) {
+		ga := action.(clientgotesting.GetAction)
+		vs := volumeSnapshot(ga.GetName(), ga.GetNamespace())
+		return true, vs, nil
+	})
+	th.fakeSnapshotClient.PrependReactor("get", "volumesnapshotcontents", func(action clientgotesting.Action) (handled bool, ret apiruntime.Object, err error) {
+		ga := action.(clientgotesting.GetAction)
+		vsc := volumeSnapshotContent(ga.GetName(), ga.GetNamespace(), th.DriverName)
+		return true, vsc, nil
+	})
+
+	th.fakeCBTClient = fakecbt.NewSimpleClientset()
+	th.fakeCBTClient.PrependReactor("get", "snapshotmetadataservices", func(action clientgotesting.Action) (handled bool, ret apiruntime.Object, err error) {
+		ga := action.(clientgotesting.GetAction)
+		if ga.GetName() != th.DriverName {
+			return false, nil, nil
+		}
+		sms := &smsv1alpha1.SnapshotMetadataService{
+			ObjectMeta: apimetav1.ObjectMeta{
+				Name: th.DriverName,
+			},
+			Spec: smsv1alpha1.SnapshotMetadataServiceSpec{
+				Audience: th.Audience,
+			},
+		}
+		return true, sms, nil
+	})
+
+	return th
+}
```

Comment:
I get what you're doing but I think this approach loses its clarity/modularity. I suggest the following steps to achieve the same functionality but with less change and only one constructor.

- Add your new fake "Clientset" variables
- Rename `RuntimeWithClientAPIs` to `Runtime` but just initialize directly with the variables (obviously initialized to `nil`), and not via function calls. i.e.
```
func (th *testHarness) Runtime() *runtime.Runtime {
	return &runtime.Runtime{
		CBTClient:      th.fakeCBTClient,
		KubeClient:     th.fakeKubeClient,
		SnapshotClient: th.fakeSnapshotClient,
		DriverName:     th.DriverName,
	}
}
```
- Rename the `FakeXXXClient` methods to be internal, and avoid the name clash with the new variables by, say, adding a prefix. e.g. `FakeKubeClient` -> `makeFakeKubeClient`.  etc.
- Then instead of making a new constructor simply make this a chainable builder method:
```
func (th *testHarness) WithFakeClientAPIs() *testHarness {
        th.fakeCBTClient = th.makeFakeCBTClient()
        th.fakeKubeClient = th.makeFakeKubeClient()
        th.fakeSnapshotClient = th.makeFakeSnapshotClient()
        return th
}
```

Usage then becomes:
```
  // construct using one of the following build chains
   th := newTestHarness()
   th := newTestHarness().WithMockCSIDriver(t)
   th := newTestHarness().WithFakeClientAPIs()
   th := newTestHarness().WithFakeClientAPIs().WithMockCSIDriver(t)
  // use
  rt := th.Runtime()
```
The pattern is then similar to the runtime test harness.